mayastor/src/core/descriptor.rs

```diff
@@ -88,6 +92,72 @@ impl Descriptor {
     pub fn into_handle(self) -> Result<BdevHandle, CoreError> {
         BdevHandle::try_from(self)
     }
+
+    /// Gain exclusive access over a block range.
+    /// The same context must be used when calling unlock.
+    pub async fn lock_lba_range(
+        &self,
+        ctx: &mut RangeContext,
+        ch: &IoChannel,
+    ) -> Result<(), nix::errno::Errno> {
+        let (s, r) = oneshot::channel::<i32>();
+        ctx.sender = Box::into_raw(Box::new(s));
+
+        unsafe {
+            let rc = bdev_lock_lba_range(
+                self.as_ptr(),
+                ch.as_ptr(),
+                ctx.offset,
+                ctx.len,
+                Some(spdk_range_cb),
+                ctx as *const _ as *mut c_void,
+            );
+            if rc != 0 {
+                return Err(nix::errno::from_i32(rc));
+            }
+        }
+
+        // Wait for the lock to complete
+        let rc = r.await.unwrap();
+        if rc != 0 {
+            return Err(nix::errno::from_i32(rc));
+        }
```

Comment:
We did try to use the ffihelper functions but unfortunately in this case that wasn't possible. For these functions SPDK expects the address of the callback argument to be the same for corresponding calls to lock and unlock. The ffihelper cb_arg function creates a new oneshot channel each time it is invoked. This was causing SPDK to fail the unlock call because the oneshot channel being used didn't match the one used by the lock call (which makes sense because it's a oneshot channel).